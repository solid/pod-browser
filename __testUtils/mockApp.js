/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {
  addStringNoLocale,
  addUrl,
  createSolidDataset,
  createThing,
  setStringNoLocale,
  setThing,
  setUrl,
} from "@inrupt/solid-client";
import { vcard, rdf, schema } from "rdf-namespaces";
import { chain } from "../src/solidClientHelpers/utils";
import { vcardExtras } from "../src/addressBook";

export const oidcPrefix = "http://www.w3.org/ns/solid/oidc#";
export const TOS_PREDICATE = `${oidcPrefix}tos_uri`;
export const POLICY_PREDICATE = `${oidcPrefix}policy_uri`;
export const CONTACTS_PREDICATE = `${oidcPrefix}contacts`;
export const LOGO_PREDICATE = `${oidcPrefix}logo_uri`;

export const APP_WEBID = "https://mockappurl.com";
export const APP_POLICY_URL = "https://mockappurl.com/privacy-policy";
export const APP_TOS_URL = "https://mockappurl.com/terms-of-service";

const contactThing = chain(
  createThing({ url: "https://mockappurl.com#contacts" }),
  (t) => addStringNoLocale(t, vcard.fn, "Example Contact"),
  (t) => addStringNoLocale(t, vcard.hasTelephone, "555-5555"),
  (t) => addStringNoLocale(t, vcard.hasEmail, "example@email.com"),
  (t) => addUrl(t, vcard.url, "https://examplewebsite.com")
);

const dataset = setThing(createSolidDataset(), contactThing);

export function mockApp() {
  return chain(
    createThing({ url: APP_WEBID }),
    (t) => addStringNoLocale(t, vcard.fn, "Mock App"),
    (t) => addUrl(t, vcardExtras("WebId"), APP_WEBID),
    (t) => addUrl(t, rdf.type, schema.SoftwareApplication), // just guessing what the type might be for apps
    (t) => addUrl(t, TOS_PREDICATE, APP_TOS_URL),
    (t) => addUrl(t, POLICY_PREDICATE, APP_POLICY_URL),
    (t) => addUrl(t, CONTACTS_PREDICATE, "https://mockappurl.com#contacts")
  );
}

export function mockAppDataset() {
  return setThing(dataset, mockApp());
}
