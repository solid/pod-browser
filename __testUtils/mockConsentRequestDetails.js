/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

async function getConsentRequestDetails() {
  return {
    "@context": [
      "https://www.w3.org/2018/credentials/v1",
      "https://w3id.org/security/suites/ed25519-2020/v1",
      "https://consent.pod.inrupt.com/credentials/v1",
    ],
    id: "https://consent.pod.inrupt.com/vc/53973727-f4d0-9e8dbdc041fd",
    type: ["VerifiableCredential", "SolidCredential", "SolidConsentRequest"],
    issuer: "https://consent.pod.inrupt.com",
    issuanceDate: "2021-05-26T16:40:03Z",
    expirationDate: "2021-06-23T16:40:03Z",
    credentialSubject: {
      id: "https://pod.inrupt.com/bob/profile/card#me",
      inbox: "https://pod.inrupt.com/bob/inbox/",
      hasConsent: {
        mode: ["Read", "Write"],
        hasStatus: "ConsentStatusRequested",
        forPersonalData: "https://pod.inrupt.com/alice/private/data",
        forPurpose: "https://example.com/SomeSpecificPurpose",
      },
    },
    proof: {
      created: "2021-05-26T16:40:03.009Z",
      proofPurpose: "assertionMethod",
      proofValue: "eqp8h_kL1DwJCpn65z-d1Arnysx6b11...jb8j0MxUCc1uDQ",
      type: "Ed25519Signature2020",
      verificationMethod: "https://consent.pod.inrupt.com/key/396f686b4ec",
    },
  };
}

export default getConsentRequestDetails;
