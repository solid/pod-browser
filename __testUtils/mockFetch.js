/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

export default function mockFetch(responses = {}) {
  return jest.fn().mockImplementation((url, init) => {
    const responseFn = responses[url.toString()];
    const responseObject = responseFn;
    const method = init?.method || "GET";
    let response;
    if (responseFn && typeof responseObject[method] !== "undefined") {
      response = responseObject[method](url, init);
    } else if (responseFn) {
      response = responseFn(url, init);
    } else {
      throw new Error(`URL (${url}) not mocked properly`);
    }
    Object.defineProperty(response, "url", { value: url });
    return Promise.resolve(response);
  });
}

export function mockSolidClientRequest(body, method = "GET", headers = {}) {
  return new Request(body, {
    headers: new Headers(headers),
    method,
  });
}

export function mockTurtleResponse(status, body = "") {
  return new Response(body, {
    status,
    headers: {
      "Content-Type": "text/turtle",
    },
  });
}
