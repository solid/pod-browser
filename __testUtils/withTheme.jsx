/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React from "react";
import { render } from "@testing-library/react";
import T from "prop-types";
import { StylesProvider, ThemeProvider } from "@inrupt/prism-react-components";
import { FeatureProvider } from "../src/contexts/featureFlagsContext";
import defaultTheme from "../src/theme";

export default function WithTheme(props) {
  const { theme, children } = props;
  return (
    <StylesProvider generateClassName={(rule) => `PodBrowser-${rule.key}`}>
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
    </StylesProvider>
  );
}

WithTheme.propTypes = {
  // eslint-disable-next-line react/forbid-prop-types
  theme: T.object,
  children: T.node.isRequired,
};

WithTheme.defaultProps = {
  theme: defaultTheme,
};

export const renderWithTheme = (children, theme = defaultTheme) => {
  return render(
    <FeatureProvider>
      <WithTheme theme={theme}>{children}</WithTheme>
    </FeatureProvider>
  );
};
