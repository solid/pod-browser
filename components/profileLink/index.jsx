/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React from "react";
import T from "prop-types";
import Link from "next/link";
import { useThing } from "@inrupt/solid-ui-react";
import { getStringNoLocale } from "@inrupt/solid-client";
import { vcard, foaf } from "rdf-namespaces";
import { useRouter } from "next/router";
import { getProfileIriFromContactThing } from "../../src/addressBook";

export const TESTCAFE_ID_PROFILE_LINK = "profile-link";

export function buildProfileLink(webId, route) {
  return `${route}/${encodeURIComponent(webId)}`;
}

export default function ProfileLink(props) {
  const { route } = useRouter();
  const { webId } = props;
  const { thing } = useThing();
  const contact = getProfileIriFromContactThing(thing);

  // Pass in an iri, or use the thing from context (such as for the contacts list)
  const profileIri = webId || contact;

  // TODO remove this once react-sdk allows property fallbacks
  const name =
    getStringNoLocale(thing, vcard.fn) ||
    getStringNoLocale(thing, foaf.name) ||
    profileIri;

  return (
    <Link href={buildProfileLink(profileIri, route)}>
      <a data-testid={TESTCAFE_ID_PROFILE_LINK}>{name}</a>
    </Link>
  );
}

ProfileLink.propTypes = {
  webId: T.string,
};

ProfileLink.defaultProps = {
  webId: null,
};
