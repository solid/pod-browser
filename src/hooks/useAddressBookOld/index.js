/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import { useEffect, useState } from "react";
import { useSession } from "@inrupt/solid-ui-react";
import useAuthenticatedProfile from "../useAuthenticatedProfile";
import { getContactsIndexIri, saveNewAddressBook } from "../../addressBook";
import { getResource } from "../../solidClientHelpers/resource";
import { ERROR_CODES, isHTTPError } from "../../error";
import useContactsContainerUrl from "../useContactsContainerUrl";

export default function useAddressBookOld() {
  const [addressBook, setAddressBook] = useState(null);
  const [error, setError] = useState(null);
  const { session } = useSession();
  const profile = useAuthenticatedProfile();
  const addressBookContainerUrl = useContactsContainerUrl();

  useEffect(() => {
    if (!session.info.isLoggedIn || !profile || !addressBookContainerUrl) {
      return;
    }
    const { webId } = profile;
    const { fetch } = session;
    const contactsIndexIri = getContactsIndexIri(addressBookContainerUrl);

    (async () => {
      const { response: existingAddressBook, error: existingError } =
        await getResource(contactsIndexIri, fetch);

      if (existingAddressBook) {
        setAddressBook(existingAddressBook.dataset);
        return;
      }

      if (existingError && isHTTPError(existingError, ERROR_CODES.NOT_FOUND)) {
        const { response: newAddressBook, error: newError } =
          await saveNewAddressBook(
            {
              iri: addressBookContainerUrl,
              owner: webId,
            },
            fetch
          );
        if (newError) {
          setError(newError);
          return;
        }
        setAddressBook(newAddressBook.index);
        return;
      }
      setError(existingError);
    })();
  }, [session, profile, addressBookContainerUrl]);

  return [addressBook, error];
}
