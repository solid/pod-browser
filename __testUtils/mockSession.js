/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import mockFetch from "./mockFetch";
import mockResponse from "./mockResponse";

import profileTurtle from "./mocks/profile.ttl";
import storageTurtle from "./mocks/podRoot.ttl";
import storageAclTurtle from "./mocks/storageAcl.ttl";

export const webIdUrl = "http://example.com/webId#me";
export const storageUrl = "http://example.com/";
export const storageAclUrl = "http://example.com/.acl";
export const anotherUsersStorageUrl = "http://anotheruser.com/";
export const anotherUsersStorageAclUrl = "http://anotheruser.com/.acl";

export { profileTurtle };

export default function mockSession(options = {}) {
  // this instance is a normal, authenticated session
  // you can override the properties with what you need, but if you repeat something a lot
  // consider naming it and create a reusable mock function
  const { webId, ...restOptions } = options;
  return {
    on: jest.fn(),
    handleIncomingRedirect: jest.fn().mockResolvedValue(null),
    fetch: mockFetch({
      [webId || webIdUrl]: () =>
        mockResponse(200, profileTurtle, { "Content-Type": "text/turtle" }),
      [storageUrl]: () =>
        mockResponse(200, storageTurtle, {
          Link: `<${storageAclUrl}>; rel="acl"`,
          "Content-Type": "text/turtle",
        }),
      [storageAclUrl]: () =>
        mockResponse(200, storageAclTurtle, { "Content-Type": "text/turtle" }),
    }),
    info: {
      isLoggedIn: true,
      sessionId: "some-session-id",
      webId: options.webId || webIdUrl,
    }, // add more properties as needed
    ...restOptions,
  }; // add more properties as needed
}

// create other functions in this file to reflect other types of sessions we need to test with

export function mockUnauthenticatedSession() {
  return {
    on: jest.fn(),
    handleIncomingRedirect: jest.fn().mockResolvedValue(null),
    fetch: () => mockFetch(),
    info: {
      isLoggedIn: false,
      sessionId: null,
      webId: null,
    },
    login: jest.fn(),
  };
}

export function mockAuthenticatedSessionWithNoAccessToPod() {
  return {
    on: jest.fn(),
    handleIncomingRedirect: jest.fn().mockResolvedValue(null),
    fetch: mockFetch({
      [webIdUrl]: () => mockResponse(200, profileTurtle),
      [storageUrl]: () =>
        mockResponse(200, "", {
          Link: `<${storageAclUrl}>; rel="acl"`,
        }),
      [storageAclUrl]: () => mockResponse(401),
    }),
    info: {
      isLoggedIn: true,
      sessionId: "some-session-id",
      webId: webIdUrl,
    },
  };
}

export function mockAuthenticatedSessionWithNoAccessToAnotherUsersPod() {
  return {
    on: jest.fn(),
    handleIncomingRedirect: jest.fn().mockResolvedValue(null),
    fetch: mockFetch({
      [webIdUrl]: () => mockResponse(200, profileTurtle),
      [anotherUsersStorageUrl]: () =>
        mockResponse(200, "", {
          Link: `<${anotherUsersStorageAclUrl}>; rel="acl"`,
        }),
      [anotherUsersStorageAclUrl]: () => mockResponse(401),
    }),
    info: {
      isLoggedIn: true,
      sessionId: "some-session-id",
      webId: webIdUrl,
    },
  };
}

export const mockAuthenticatedSession =
  mockAuthenticatedSessionWithNoAccessToPod;
