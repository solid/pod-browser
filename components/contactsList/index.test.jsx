/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React from "react";
import * as solidClientFns from "@inrupt/solid-client";
import { useRouter } from "next/router";
import { foaf, schema } from "rdf-namespaces";
import { screen, waitFor } from "@testing-library/react";
import * as addressBookFns from "../../src/addressBook";
import useAddressBookOld from "../../src/hooks/useAddressBookOld";
import useContactsOld from "../../src/hooks/useContactsOld";
import useFullProfile from "../../src/hooks/useFullProfile";
import useProfiles from "../../src/hooks/useProfiles";
import { renderWithTheme } from "../../__testUtils/withTheme";
import ContactsList, { handleDeleteContact } from "./index";
import {
  aliceWebIdUrl,
  bobWebIdUrl,
  mockPersonDatasetAlice,
  mockPersonThingAlice,
  mockPersonThingBob,
} from "../../__testUtils/mockPersonResource";
import mockSession from "../../__testUtils/mockSession";
import mockSessionContextProvider from "../../__testUtils/mockSessionContextProvider";
import { TESTCAFE_ID_PROFILE_LINK } from "../profileLink";

jest.mock("../../src/hooks/useAddressBookOld");
jest.mock("../../src/hooks/useContactsOld");
jest.mock("../../src/hooks/useProfiles");
jest.mock("../../src/hooks/useFullProfile");
jest.mock("next/router");

const mockedUseRouter = useRouter;

describe("ContactsList", () => {
  beforeEach(() => {
    mockedUseRouter.mockReturnValue({
      route: "/contacts",
    });
  });
  const session = mockSession();
  const profile = { webIdProfile: mockPersonDatasetAlice() };
  const SessionProvider = mockSessionContextProvider(session, false, profile);
  it("renders spinner while useAddressBookOld is loading", () => {
    useAddressBookOld.mockReturnValue([null, null]);
    useContactsOld.mockReturnValue({
      data: undefined,
      error: undefined,
      mutate: () => {},
    });
    useProfiles.mockReturnValue(null);

    const { asFragment } = renderWithTheme(
      <SessionProvider>
        <ContactsList />
      </SessionProvider>
    );
    expect(asFragment()).toMatchSnapshot();
    expect(useAddressBookOld).toHaveBeenCalledWith();
    expect(useContactsOld).toHaveBeenCalledWith(null, foaf.Person);
  });

  it("renders spinner while useContactsOld is loading", () => {
    useAddressBookOld.mockReturnValue([42, null]);
    useContactsOld.mockReturnValue({
      data: undefined,
      error: undefined,
      mutate: () => {},
    });
    useProfiles.mockReturnValue(null);

    const { asFragment } = renderWithTheme(
      <SessionProvider>
        <ContactsList />
      </SessionProvider>
    );
    expect(asFragment()).toMatchSnapshot();
    expect(useContactsOld).toHaveBeenCalledWith(42, foaf.Person);
  });

  it("renders spinner while useProfiles is loading", () => {
    useAddressBookOld.mockReturnValue([42, null]);
    useContactsOld.mockReturnValue({
      data: "peopleData",
      error: undefined,
      mutate: () => {},
    });
    useProfiles.mockReturnValue(null);

    const { asFragment } = renderWithTheme(
      <SessionProvider>
        <ContactsList />
      </SessionProvider>
    );
    expect(asFragment()).toMatchSnapshot();
    expect(useProfiles).toHaveBeenCalledWith("peopleData");
  });

  it("renders error if useAddressBookOld returns error", () => {
    useAddressBookOld.mockReturnValue([null, "error"]);
    useContactsOld.mockReturnValue({
      data: undefined,
      error: undefined,
      mutate: () => {},
    });

    const { asFragment } = renderWithTheme(
      <SessionProvider>
        <ContactsList />
      </SessionProvider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders page when people is loaded", async () => {
    useFullProfile
      .mockReturnValueOnce({
        webId: aliceWebIdUrl,
        names: ["Alice"],
        types: [schema.Person],
      })
      .mockReturnValueOnce({
        webId: bobWebIdUrl,
        names: ["Bob"],
        types: [schema.Person],
      });
    useAddressBookOld.mockReturnValue([42, null]);
    useContactsOld.mockReturnValue({
      data: "peopleData",
      error: undefined,
      mutate: () => {},
    });
    useProfiles.mockReturnValue([mockPersonThingAlice(), mockPersonThingBob()]);
    const { asFragment, getAllByTestId } = renderWithTheme(
      <SessionProvider>
        <ContactsList />
      </SessionProvider>
    );

    await waitFor(() => {
      expect(getAllByTestId(TESTCAFE_ID_PROFILE_LINK)[0]).toHaveAttribute(
        "href",
        `/contacts/${encodeURIComponent(aliceWebIdUrl)}`
      );
      expect(getAllByTestId(TESTCAFE_ID_PROFILE_LINK)[1]).toHaveAttribute(
        "href",
        `/contacts/${encodeURIComponent(bobWebIdUrl)}`
      );
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders webId for failed profiles and name for successfully fetched profiles", async () => {
    useAddressBookOld.mockReturnValue([42, null]);
    const webId = "https://somewebid.com";
    const contact = solidClientFns.addUrl(
      solidClientFns.createThing(),
      addressBookFns.vcardExtras("WebId"),
      webId
    );
    useContactsOld.mockReturnValue({
      data: "peopleData",
      error: undefined,
      mutate: () => {},
    });
    useProfiles.mockReturnValue([mockPersonThingAlice(), contact]);
    useFullProfile
      .mockReturnValueOnce({
        webId: aliceWebIdUrl,
        names: ["Alice"],
        types: [schema.Person],
      })
      .mockReturnValueOnce({
        webId,
        names: [],
        types: [schema.Person],
      });

    const { asFragment, getAllByTestId } = renderWithTheme(
      <SessionProvider>
        <ContactsList />
      </SessionProvider>
    );
    await waitFor(() => {
      expect(getAllByTestId(TESTCAFE_ID_PROFILE_LINK)[0]).toHaveTextContent(
        "Alice"
      );
      expect(getAllByTestId(TESTCAFE_ID_PROFILE_LINK)[1]).toHaveTextContent(
        webId
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });

  it("renders empty state message when there are no contacts", () => {
    useAddressBookOld.mockReturnValue([42, null]);
    useContactsOld.mockReturnValue({
      data: "peopleData",
      error: undefined,
      mutate: () => {},
    });
    useProfiles.mockReturnValue([]);

    renderWithTheme(
      <SessionProvider>
        <ContactsList />
      </SessionProvider>
    );

    const message = screen.getByText("You donâ€™t have any contacts yet!");

    expect(message).toBeTruthy();
  });

  it("renders error if useContacts returns error", () => {
    useAddressBookOld.mockReturnValue([42, null]);
    useContactsOld.mockReturnValue({
      data: undefined,
      error: "error",
      mutate: () => {},
    });

    const { asFragment } = renderWithTheme(
      <SessionProvider>
        <ContactsList />
      </SessionProvider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

describe("handleDeleteContact", () => {
  it("returns a handler that deletes a contact, updates people data and closes drawer", async () => {
    const mockDeleteContact = jest
      .spyOn(addressBookFns, "deleteContact")
      .mockImplementation(() => null);
    const addressBookUrl = "http://example.com/contacts";
    const contact = mockPersonDatasetAlice();
    const addressBook = "address book";
    const closeDrawer = jest.fn();
    const fetch = jest.fn();
    const people = [contact];
    const peopleMutate = jest.fn();
    const selectedContactIndex = 0;

    jest.spyOn(solidClientFns, "getSourceUrl").mockReturnValue(addressBookUrl);

    const handler = handleDeleteContact({
      addressBook,
      closeDrawer,
      fetch,
      people,
      peopleMutate,
      selectedContactIndex,
    });

    await handler();

    expect(mockDeleteContact).toHaveBeenCalledWith(
      addressBookUrl,
      contact,
      foaf.Person,
      fetch
    );
    expect(peopleMutate).toHaveBeenCalled();
    expect(closeDrawer).toHaveBeenCalled();
  });
});
