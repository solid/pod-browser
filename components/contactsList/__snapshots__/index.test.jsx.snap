// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactsList renders error if useAddressBookOld returns error 1`] = `
<DocumentFragment>
  error
</DocumentFragment>
`;

exports[`ContactsList renders error if useContacts returns error 1`] = `
<DocumentFragment>
  error
</DocumentFragment>
`;

exports[`ContactsList renders spinner while useAddressBookOld is loading 1`] = `
<DocumentFragment>
  <div
    class="PodBrowser-loading-indicator-container PodBrowser-loading-indicator-container--center"
  >
    <svg
      class="PodBrowser-loading-indicator"
      data-testid="spinner"
      viewBox="22 22 44 44"
    >
      <circle
        cx="44"
        cy="44"
        fill="none"
        r="20.2"
        stroke-width="3.6"
      />
    </svg>
  </div>
</DocumentFragment>
`;

exports[`ContactsList renders spinner while useContactsOld is loading 1`] = `
<DocumentFragment>
  <div
    class="PodBrowser-loading-indicator-container PodBrowser-loading-indicator-container--center"
  >
    <svg
      class="PodBrowser-loading-indicator"
      data-testid="spinner"
      viewBox="22 22 44 44"
    >
      <circle
        cx="44"
        cy="44"
        fill="none"
        r="20.2"
        stroke-width="3.6"
      />
    </svg>
  </div>
</DocumentFragment>
`;

exports[`ContactsList renders spinner while useProfiles is loading 1`] = `
<DocumentFragment>
  <div
    class="PodBrowser-loading-indicator-container PodBrowser-loading-indicator-container--center"
  >
    <svg
      class="PodBrowser-loading-indicator"
      data-testid="spinner"
      viewBox="22 22 44 44"
    >
      <circle
        cx="44"
        cy="44"
        fill="none"
        r="20.2"
        stroke-width="3.6"
      />
    </svg>
  </div>
</DocumentFragment>
`;
